### DISCUSSION TITLE
How to specify a GPU from all GPUs

### DISCUSSION THREAD

**Discussion by biglinn** [deepmodeling/dpgen#1664] [Q&A]
Hi,

I'm going to trian deepmd-kit models by DPGEN workflow. And I have a doubt.
When there are multiple GPUs on the server, Could I specify a specific GPU in the `machine.json`?

**Comment by biglinn** [comment]
If the job directly runs at the workstation with multiple GPUs, will there be a situation where **different GPUs repeatedly run the job**?

Because I find there is repetitive information in the `dpgen.log`, like this:

![image](https://github.com/user-attachments/assets/4ae599c1-9dc9-4fee-bcd9-b1a788351e22)

**Comment by njzjz** [comment]
The environment variables `CUDA_VISIBLE_DEVICES` can be used. There is a lot of information on this environment variable on the Internet.

For setting environment variables in the jobs, refer to https://docs.deepmodeling.com/projects/dpdispatcher/en/latest/resources.html#argument:resources/envs

**Reply by biglinn** [reply]
I follow the script in https://docs.deepmodeling.com/projects/dpdispatcher/en/latest/examples/shell.html
Like this:
```
"strategy": {
    "if_cuda_multi_devices": true
},
"env": {
    "CUDA_VISIBLE_DEVICES": 2
},
```
The `if_cuda_multi_devices` of `strategy` really works. This command submits all jobs to only one GPU.
But `"CUDA_VISIBLE_DEVICES": 2` doesn't work. It can't specify which GPU to submit jobs to, while all jobs are submitted to the first GPU by default. 
Is there other problem?

**Reply by njzjz** [reply]
Do you use them together? They might provide conflict `CUDA_VISIBLE_DEVICES`.

**Reply by biglinn** [reply]
I used them together, while I also used them **respectively**:
like this:
```
"strategy": {
    "if_cuda_multi_devices": true
},
```
and this:
```
"env": {
    "CUDA_VISIBLE_DEVICES": 2
},
```
Only the `if_cuda_multi_devices` dose work.

**Reply by biglinn** [reply]
Oh, I find this variable is shown as unspecified in the train.log:
![image](https://github.com/user-attachments/assets/594c23d5-2880-476f-9431-c6b4aba09055)

Is this setting not valid?
```
"env": {
    "CUDA_VISIBLE_DEVICES": 2
},
```

**Reply by njzjz** [reply]
It's `envs` instead of `env`.

**Reply by biglinn** [reply]
Oh, it is! 
Haha, how stupid of me to make this mistake.
